include/deal.II/fe/fe_poly.templates.h

```diff
@@ -223,6 +224,56 @@ FE_Poly<PolynomialType, dim, spacedim>::requires_update_flags(
 //---------------------------------------------------------------------------
 
 
+
+/**
+ * Returns whether we need to correct the Hessians and third derivatives with
+ * the derivatives of the Jacobian. This is determined by checking if
+ * the jacobian_pushed_forward are zero.
+ *
+ * Especially for the third derivatives, the correction term is very expensive,
+ * which is why we check if the derivatives are zero before computing the
+ * correction.
+ */
+template <int dim, int spacedim>
+bool
+higher_derivatives_need_correcting(
+  const Mapping<dim, spacedim> &mapping,
+  const internal::FEValuesImplementation::MappingRelatedData<dim, spacedim>
+    &                mapping_data,
+  const unsigned int n_q_points,
+  const UpdateFlags  update_flags)
+{
+  // If higher derivatives weren't requested we don't need to correct them.
+  const bool update_higher_derivatives =
+    (update_flags & update_hessians) || (update_flags & update_3rd_derivatives);
+  if (!update_higher_derivatives)
+    return false;
+
+  // If we have a Cartesian mapping, we know that jacoban_pushed_forward_grads
+  // are identically zero.
+  if (dynamic_cast<const MappingCartesian<dim> *>(&mapping))
+    return false;
+
+  // Go through all entries in jacobian_pushed_forward. If we find a non-zero
+  // value, we know that the higher derivatives need to be corrected.
+  for (unsigned int q = 0; q < n_q_points; ++q)
+    for (unsigned int i = 0; i < spacedim; ++i)
+      for (unsigned int j = 0; j < spacedim; ++j)
+        for (unsigned int k = 0; k < spacedim; ++k)
+          {
+            const double jacobian_pushed_forward_grad =
+              mapping_data.jacobian_pushed_forward_grads[q][i][j][k];
+
+            if (std::abs(jacobian_pushed_forward_grad) > 0)
```

Comment:
Is this what you had in mind @bangerth ?

I guess the `jacobian_pushed_forward_grads` will not be identically zero if they are computed, right?
So I should probably check
`if (std::abs(jacobian_pushed_forward_grad) > tol)` 
here, but I don't really know what a good value of `tol` would be. :thinking:

Reply:
For some mappings, they will be exactly zero -- for example for the `MappingCartesian`. For others that's maybe not the case, but you can think about what these third derivatives represent: derivatives of second derivatives. So you can say that the third derivative 
```
  f'''(x) = lim_{eps->0}  [f''(x+eps)-f''(x)]/eps
```
is essentially zero if
```
  |f'''(x)|  << |f''(x)]| / L
```
where `L` is some sort of length scale -- for example the mesh size. In other words, here you would want to test something along the lines of
```
  if (mapping.jacobian_pushed_forward_grad[q].norm() 
      <=
      1e-12 * mapping.jacobian_pushed_forward[q].norm() / cell->diameter())
    ... term is essentially zero ...
```
A safer bet would probably to to take the max over all `mapping.jacobian_pushed_forward[q]` on the right hand side, just to make sure that you capture the case where by chance the `mapping.jacobian_pushed_forward[q]` at one point is small.

Makes sense?

Reply:
How is the state of this discussion? In case there is not an immediate solution in terms of the tolerances, I suggest to merge this PR as is (given that it represents already a good improvement for the cases @simonsticko is interested in) and leave the rest to an issue.

Reply:
That would be fine with me.

@simonsticko ?

Reply:
Well, I'm mainly interested in the case where the mapping in Cartesian. So should I just remove the added check on the values of `jacobian_pushed_forward_grads`?

Reply:
We would be fine with merging this and maybe improve later if you agree.

Reply:
I'm okey with merging this. However, if this:
`std::abs(jacobian_pushed_forward_grad) > 0`
is always `true` in practice (due to floating-point-computations) I suspect that these changes might slightly slow down the general case (when MappingCartesian isn't used).

Reply:
I see. I think just commenting the code above and returning `true` at this point is OK for now.

Reply:
Okey, I have added a comment and added issue #9682.